### ISSUE TITLE
Guiding Speculation

### ISSUE THREAD

**Issue by RossTate** [WebAssembly/gc#251]
If we're exploring speculation now, it would be good to provide the generator with a means to guide speculation. This would both help make better use of speculation and help make the effect of speculation more predictable.

At present, it sounds like speculation is being done at the call-site level. But other VMs do speculation at the "class" level. This has the effect of grouping a bunch of call sites (e.g. calls to methods on the same receiver) and propagating speculation into inlined calls (that is, if one inlines a method implementation that calls a method on `this`, then that method can be devirtualized rather than have to respeculated).

One way we could support this is by adding an instruction like `guarded_inline $local $field num instr* end`. This instruction informs the engine to make a guarded inline cache with `num` cases. If the value of `struct.get $field (local.get $local)` matches one of these cases, then the corresponding optimized code is jumped to. That optimized code is `instr*` compiled with the assumption that `struct.get $field (local.get $local)` returns that value (where `$field` must be immutable, and assuming `$local` doesn't change within `instr*`).

For example, if `$field` is a v-table, then this code will be optimized for that particular v-table. In particular, the results of any loads (say of `funcref`s) from immutable fields of this v-table can be inlined, as can any calls to those loaded `funcref`s. Similarly, if i-tables are made immutable, then the result of the entire search through the i-table can be inlined, as can the call on the retrieved `funcref`. This recreates the "class"-level speculation used by other VMs.

(If we go with the nominal approach and furthermore add direct support for adding fields&mdash;such as v-table `funcref`s&mdash;to the object descriptor of a given nominal type, then this becomes even easier to guide.)

**Comment by jakobkummerow** [comment]
> If we're exploring speculation now

I think this premise alone deserves quite a bit of debate. I see speculation first and foremost as an engine-internal implementation detail. Engines are free to do it, or not do it, or do it sometimes, or whatever. I think standardizing a programmable (i.e. controllable by the module) speculation or inline-caching system (which are two distinct things, btw) in engines is waaay out of scope for WasmGC MVP; and it remains to be seen/discussed whether we'll ever want to do that.

**Comment by kripken** [comment]
Makes sense @titzer , thanks. Do you perhaps have a reference to a summary of the state of the art in this area? Basically what I have not been able to get is a clear answer to how good can we expect an AOT compiler of Java to actually do. In particular I'd be curious about production AOT compilers, like Graal.

For comparison, I am aware of AOT compilers for C# that at least at one point were considered fairly fast compared to dynamic VMs ([il2cpp vs Mono](https://docs.unity3d.com/Manual/IL2CPP.html)).

I definitely agree a single benchmark is not enough to focus on - I'd be curious to hear more about the benchmarking you were doing @RossTate ?

**Comment by RossTate** [comment]
> For comparison, I am aware of (edit: production) AOT compilers for C# that at least at one point were considered fairly fast compared to dynamic VMs (il2cpp vs Mono).

My understanding is that even C# VMs do not make much use of speculation or closed-world assumptions. The jitting feature is primarily for faster loading, supporting dynamic loading, and supporting dynamic code generation. So the compilation model is not far from WebAssembly's. I believe the major difference is that compilation is done during instantiation, which means it doesn't have to deal with instance objects, and means that the compiler can incorporate the specific values used at instantiation (e.g. inlining imported functions or specializing to particular constants/types).

**Comment by kripken** [comment]
Thanks @RossTate ! Very interesting, I'll be taking a close look at all that.

Meanwhile I've been looking for data on production AOT compilers for Java, and the best I can find is this 2019 talk on Graal

https://qconsf.com/system/files/presentation-slides/qconsf2019-alina-yurenko-jit-vs-aot-performance-with-graalvm.pdf

Slide 22 has an Apache benchmark which makes the argument that PGO is needed for AOT to match JIT (or else it loses some 20% or so). Overall the talk suggests AOT is best for startup and JIT for peak performance.

**Comment by RossTate** [comment]
Oh, nice find! I wish I had suggestions to help you on your search.

One more thing I found is that their AOT compiler significantly outperforms HotSpot's JIT for *Scala* programs, even without profiling. [This article](https://medium.com/graalvm/compiling-scala-faster-with-graalvm-86c5c0857fa3) it refers to seems to suggest that this is due to HotSpot's JIT being specifically tuned for common patterns in Java that Scala programs do not exhibit as often (despite compiling to the same VM), possibly even hurting performance. Their "enterprise" AOT compiler (without profiling information) outperforms HotSpot's JIT by more than 40%, despite seemingly not being designed for specifically Scala.

My sense from that is that AOT can provide overall good performance in a general-purpose manner, which JIT can improve upon when properly guided. It sounds like that room for improvement when properly guided is plausibly around 20% (which coincidentally is how much speedup we observed when we fully devirtualized the benchmark discussed above). (This is in a closed-world setting. In an open-world setting, which would be nice to support efficiently in a Post-MVP, there is likely many more opportunities for JIT to improve performance even with less guiding.)
